Issue/PR Templates

rmainwork opened the pull request:
I closed https://github.com/seas-computing/course-planner/pull/12 and re-opened the PR here instead. This feels like a more suitable place for this work.

@jonseitz - I made some of the changes you requested in the previous PR. In particular, you mentioned priority and related issues. Is [this](https://github.com/seas-computing/react-boilerplate/commit/b178c261e39b3fee35780f45ce432921a8f03aba#diff-195a635ad245ded9076330e31134bd80R19) what you had in mind?

Commit:
User Story Template

Add issue template to create "As a ___ I want to ___ so that ___" type user stories. Sadly, this does not seem to work with project cards, but it's a good starting point for issues

Commit:
Bug Report Template

Add simple template for reporting bugs with questions to solicit helpful triage information

Commit:
Pull Request Template

Add PR template to elicit helpful information from developers when opening a PR

jonseitz commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -0,0 +1,34 @@
+<!--
+  Go over all the following points, and put an `x` in all the boxes that apply.
+-->
+- [ ] I run eslint on the code
```
should be: I **have** run

rmainseas replied:
Whoops! Fixed in 1778b5c45bb3c074d3245de4b1955f0beac10a11

rmainwork replied:
Whoops! Fixed in 1778b5c45bb3c074d3245de4b1955f0beac10a11

jonseitz commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -0,0 +1,34 @@
+- [ ] I have added tests to cover my changes.
+
+## Priority:
+- [ ] Normal <!-- New piece of functionality -->
```
Might be worth having a **low** option as well for non-essential changes

rmainwork replied:
I wondered about that, and actually put it in at first, but changed my mind since PRs are supposed to be high priority generally (for review)...if you see what I mean. I was cautious of suggesting that any PR is low priority...

jonseitz replied:
That's a fair point. @chrxr should probably weigh in on this, but think there are cases where we may have a low-priority change (e.g. tweaking minor config files). I guess I think about it in terms of "how soon does this need to be reviewed?", along the lines of:

**high** drop everything and review this now
**normal** finish what you're doing and review this as soon as you can
**low** review this sometime this week

jonseitz commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -0,0 +1,34 @@
+  Mention any issues that this PR is meant to address or is related to below.
+  If possible, please provide more information regarding the relationship to
+  the issues in question (i.e does the PR fix the issue, is it designed to fix
+  another bug that is similar to the issue etc.)
```
I'd suggest making an associated issue number mandatory, and maybe adding something like `fixes #__` to make use of the issue auto-close feature.

rmainwork replied:
Yeah, that's not a bad idea. Though obviously the `___` is just there to make the gaps easy to select. It doesn't actually do anything fancy with GitHub.

rmainwork replied:
Added this in 7f30f2aad2faa57625ffde372a6af816d708f3ca

jonseitz reviewed (changes requested):
Few more notes below. 

I also still think we should have some sort of "development task" issue template based on the kanban cards we used in the EAF project. If nothing else, it could at least start a features checklist with the last three items we always have:

```md
- [ ] 
- [ ] test
- [ ] docs
- [ ] review
```

I suspect we'll be the primary users of these templates, so we might as them as specific to our needs as possible.

Commit:
Fix grammar for eslint checkbox

rmainwork commented:
The only thing I would say about the test/docs/review items is that I feel like they might now be superceded by the kanban. That is to say that we used the `review` box to illustrate when the review had been approved vs was still under review....and I would say that maybe the tests and docs are perhaps just part of the development phase. Not least because some issues/PRs may not require docs....I'm willing to have my mind changed though.

Commit:
PR Issue Number

Add issue number requirement in to PR template to encourage people to note the issue that their PR is meant to address.

jonseitz commented:
I agree, it probably makes more sense to just have the kanban handle that tracking for us. Though that same reasoning could apply to the check boxes in the PR template for linting and testing -- ideally Travis/codecov are going to be monitoring that for us and provide some kind of warning if those aren't done.

I guess my main concern here is that we want make sure we're creating templates that match our development process and encourage good habits. Looking at the EAF project, most of the issues we created were checklists of what needed to be added for a feature, so I think it makes sense to have an issue template with a checklist already started so @chrxr doesn't have to type it live. Maybe it makes more sense to just have three blank checkboxes to start it off?

rmainwork commented:
Fair enough...or maybe one blank box and then the rest can be added if need be?

rmainwork commented:
After discussion with @chrxr - I'll put the "Tests" and "Docs" checkboxes in and omit the "review" one. Sound good?

Commit:
Checkboxes

Add "Tests" and "Docs" checkboxes to templates to denote progress

jonseitz commented:
Yep, that sounds good to me

patking02 reviewed (approved):
Looks great; thanks.
